### PR TITLE
[DEM] New Constitutive Laws and new Element to deal with them

### PR DESCRIPTION
**Pull request by joaquinirazabal**
Two new constitutive laws have been created:

1) DEM_D_Stress_dependent_cohesive_CL: a cohesive law where the cohesion depends on the history of contact forces. It was taken from https://www.sciencedirect.com/science/article/abs/pii/S0022489814000159.

2) DEM_D_Conical_damage_CL: a contact law where the contacting area changes depending on the history of contact forces taken from https://bit.ly/2HZI07t. This CL replaces DEM_D_Hertz_depenedent_friciton_CL as it is an improved one.

As those constitutive laws need to store a lot of information from the contacts I created a new element derived from SphericParticle called ContactInfoSphericParticle to deal with that information.

### CODE REVIEW

**Comment by farrufat-cimne** [comment]
Seems good. I didnt see any change related to the interface tho. How do you choose to use the ContactInfoSphericParticle3D? Is it implicit to the cohesive stress dependent law?

thx @joaquinirazabal

**Comment by joaquinirazabal** [comment]
The Element type "ContactInfoSphericParticle3D" should be specify in the mdpa file. There are also other changes that should be made in the interface.

Yesterday I forgot to mention that my plan is including those changes in the interface in another PR in the following days (as this does not break anything).

Now I have a doubt, should I implement the changes in the interface only in the new interface or in both the new and the old one?

thx @farrufat-cimne

**Comment by farrufat-cimne** [comment]
@joaquinirazabal , I would say its better to wait a bit before implementing it in the interface.
Since we are moving to the new one, it would be a waste of time implementing it in the old interface:
- The element type was selected using set ElementType [::wkcf::GetElementType] that is based on the ActiveAppList to select the corresponding element type as can be seen in : 
if {"Fluid" in $ActiveAppList} { ....... set gelemlist [list "SwimmingDEMElement"]

The new interface do not use anything similar to ActiveAppList. Instead the element comes implicit in the app (as far as I can see) so right now I'm not sure how would it be done:
in parts.spd of the new DEM interface:
```

<condition n="Parts" pn="Parts" ov="point,line,surface,volume" ovm="element" icon="shells16" help="Select your group" un="DEMParts">
	<value n="Element" pn="Element" actualize="1" values="" v="" dict="[GetElements ElementType DEM]" help="Select an element" state="hidden">
		<dependencies node="../value[@n = 'ConstitutiveLaw']" actualize="1" />
	</value>
```

The ElementType comes from Elements.xml defined as:
```

<ElementList>
  <ElementItem n="SphericPartDEMElement3D" pn="Spheric Element" ImplementedInFile=".cpp" ImplementedInApplication="DEMApplication" ProductionReady="ProductionReady" WorkingSpaceDimension="3D" ElementType="DEM" help="">
    <!--here we could add a list of all of the possible geometries-->
    <TopologyFeatures>
      <item GeometryType="Sphere" nodes="1" KratosName="SphericParticle3D"/>
    </TopologyFeatures>

```

File: applications/DEMApplication/custom_constitutive/DEM_D_DMT_cohesive_law.cpp [hunk 1]
```diff
@@ -32,17 +32,17 @@ namespace Kratos {
         const double radius_sum     = my_radius + other_radius;
         const double radius_sum_inv = 1.0 / radius_sum;
         const double equiv_radius   = my_radius * other_radius * radius_sum_inv;
-        const double cohesive_force = 8.0 * Globals::Pi * equiv_cohesion * equiv_radius;
+        const double cohesive_force = 2.0 * Globals::Pi * equiv_cohesion * equiv_radius;
```
Comment (maceligueta) [inline applications/DEMApplication/custom_constitutive/DEM_D_DMT_cohesive_law.cpp:35]: @salvalatorre : you coded this. Do you remember why 8.0? 
@joaquinirazabal : any clarification is welcome.

File: applications/DEMApplication/custom_constitutive/DEM_D_JKR_cohesive_law.cpp [hunk 2]
```diff
@@ -38,13 +38,13 @@ namespace Kratos {
         const double radius_sum_inv = 1.0 / radius_sum;
         const double equiv_radius   = my_radius * other_radius * radius_sum_inv;
         const double contact_radius = sqrt(equiv_radius * indentation);
-        const double cohesive_force = equiv_young * sqrt(8.0 * equiv_cohesion * Globals::Pi * contact_radius * contact_radius * contact_radius / equiv_young);
+        const double cohesive_force = sqrt(8.0 * equiv_cohesion * Globals::Pi * contact_radius * contact_radius * contact_radius * equiv_young);
```
Comment (maceligueta) [inline applications/DEMApplication/custom_constitutive/DEM_D_JKR_cohesive_law.cpp:41]: Wow! This is a big change/bugfix!!

Comment (farrufat-cimne) [reply applications/DEMApplication/custom_constitutive/DEM_D_JKR_cohesive_law.cpp:41]: so the first review on  const double cohesive_force ~  2*Pi*r
in the second review  const double cohesive_force ~  8*Pi*r^3
correct?

File: applications/DEMApplication/custom_constitutive/DEM_D_DMT_cohesive_law.cpp (hunk 1 above)
Comment (joaquinirazabal) [reply applications/DEMApplication/custom_constitutive/DEM_D_DMT_cohesive_law.cpp:35]: I think that 8.0 came from the JKR_CL but in the DMT_CL it should be 2.0,

**Review by farrufat-cimne — APPROVED** [review]
I guess it can be merged.

**Pull request merged**